{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Projet 1 : Alignement des séquences et programmation dynamique\n",
    "BAKKALI Yahya : 000445166"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "En bio-informatique, l'alignement de séquences est une manière de représenter deux ou plusieurs séquences de macromolécules biologiques (ADN, ARN ou protéines) les unes sous les autres, de manière à en faire ressortir les régions homologues ou similaires [1]. Dans le cadre de ce projet on étudiera seulement l'alignement par paires et comment réaliser un alignement optimal entre deux séquences de protéines en utilisant la programmation dynamique comme méthode parmi les autres méthodes existantes. La programmation dynamique consiste à résoudre un problème en le décomposant en sous-problèmes, puis à résoudre les sous-problèmes, des plus petits aux plus grands en stockant les résultats intermédiaires [2]. La visualisation des régions de similarités obtenues par l'alignement des séquences permet d'identifier si elles sont la conséquence de la même fonctionnalité ou de la même structure de domaine ou plutôt une relation évolutive entre les deux séquences. Il y en a deux types d'alignement un global qui se restreint sur la totalité de la longueur des deux séquences et l'autre local qui se focalise seulement sur des segments limités de la séquence où le taux de similarité est fort par rapport à la totalité de la séquence. Parmi les algorithmes réalisés pour obtenir les alignements optimaux on trouve deux trop connue celui de Needleman-Wunsch qui réalise l'alignement global et l'autre de Smith-Waterman qui réalise l'alignement local. Le meilleur alignement des séquences correspond au meilleur score possible de correspondance entre deux séquences, les matrices de substitution permettent de calculer ce score. Les matrices de substitution sont des matrices utilisées en bioinformatique pour réaliser des alignements de séquences biologiques reliées évolutivement. Elles permettent de donner un score de similarité ou de ressemblance entre deux acides aminés [3].Il existe plusieurs de ces matrices, basées sur des principes de construction différents. Par exemple la matrice __PAM__ est base sur les mutations observées dans un alignement global avec la conservation des régions mutables tandis que la matrice __BLOSUM__ n'est fondée que sur des segments spécifiques dans la séquence sans l'introduction des trous. Le numéro de PAM indique la divergence entre les séquences et le numéro de BLOSUM indique le pourcentage d'identité."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Matériel & Méthodes\n",
    "Pour la réalisation de ce projet on utilisera Python3 comme langage de programmation et des jeux données pour tester et évaluer les algorithmes d'alignement :\n",
    "1. Deux types de matrices de substitution __PAM__ et __BLOSUM__ stockée dans des fichiers textes blosum62, blosum80,pam60,pam120.\n",
    "2. Deux séquences au format FASTA __WW-squence.fasta__ pour le global alignement et __protein-sequences.fasta__ pour le local alignement. \n",
    "\n",
    "Avant l'extraction des données à partir des fichiers fournies on a besoin du certaines structures pour les stocker.<br> Pour cela on va créer deux ADT(Abstract Data Type) un pour la séquence des acides amines et l'autre pour les matrices de scores et de substitutions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ADTsequence(object):\n",
    "\n",
    "    def __init__(self, sequence):\n",
    "        self.sequence = sequence\n",
    "\n",
    "    def __getitem__(self, key):\n",
    "        return self.sequence[key]\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sequence)\n",
    "\n",
    "    def __str__(self):\n",
    "        output = \"\"\n",
    "        for i in range(len(self) // 60 + 1):\n",
    "            output += self.sequence[i * 60:(i + 1) * 60]\n",
    "            output += \"\\n\"\n",
    "        return output[:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ADTsequence est un ADT qui représente une séquence d’acides aminés et toutes les opérations qu’on peut exécuter sur une séquence qui sont nécessaires pour l’implémentation d’algorithme d’alignement (p.ex. retourner un acide aminé à une certaine position ou visualiser la séquence en format FASTA)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ADTmatrix:\n",
    "\n",
    "    def __init__(self, n=None, m=None):\n",
    "        \n",
    "        self.n = n\n",
    "        self.m = m\n",
    "\n",
    "        if n and m:\n",
    "            self.matrix = [[0 for i in range(m)] for j in range(n)]\n",
    "        else:\n",
    "            self.matrix = []\n",
    "\n",
    "    def getItem(self, i, j):\n",
    "        return self.matrix[i][j]\n",
    "\n",
    "    def setItem(self, i, j, value):\n",
    "        self.matrix[i][j] = value\n",
    "\n",
    "    def maxValue(self):\n",
    "\n",
    "        x, y = 1, 1\n",
    "        maximum = self.getItem(x, y)\n",
    "        for i in range(1, self.n):\n",
    "            for j in range(1, self.m):\n",
    "                if maximum < self.getItem(i, j):\n",
    "                    maximum = self.getItem(i, j)\n",
    "                    x, y = i, j\n",
    "        return x, y\n",
    "        \n",
    "    def addRow(self, row):\n",
    "        self.matrix.append(row)\n",
    "\n",
    "    def __str__(self):\n",
    "        output = \"\"\n",
    "        for i in range(len(self.matrix)):\n",
    "            for j in range(len(self.matrix[0])):\n",
    "                output += str(self.matrix[i][j])\n",
    "                output += \" \"\n",
    "            output += \"\\n\"\n",
    "        return output[:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ADTmatrix est un ADT qui représente une matrice et les opérations qu’on peut exécuter sur cette matrice.Il sera utiliser pour créer des sous-classes matrice, pour la matrice de substitution et pour les matrices scores que nous utiliserons lors de l’implémentation de l’algorithme d’alignement de séquences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SubstitutionMatrix(ADTmatrix):\n",
    "\n",
    "    def __init__(self, n=None, m=None):\n",
    "        super().__init__(n, m)\n",
    "        self.myList = []\n",
    "\n",
    "    def __getitem__(self, key):\n",
    "        return self.myList.index(key)\n",
    "\n",
    "    def Append(self, value):\n",
    "        self.myList.append(value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SubstitutionMatrix est une sous-classe de la classe ADTmatrix dans laquelle on stockera les matrices de substitution qui seront utilisé après dans les algorithmes d'alignement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SequencesParser(filename):\n",
    "    f = open(filename, encoding='utf-8')\n",
    "    mylist = []\n",
    "    sequence = ''\n",
    "    for line in f:\n",
    "        if line[0] not in [\">\", \" \", \"\\n\"]:\n",
    "            sequence += line.strip(\"\\n\")\n",
    "        elif len(sequence) :\n",
    "            mylist.append(ADTsequence(sequence))\n",
    "            sequence = ''\n",
    "    if len(sequence) :\n",
    "        mylist.append(ADTsequence(sequence))\n",
    "    f.close()\n",
    "    return mylist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "sequenceParser un parser qui peut lire des fichiers avec des séquences (en format FASTA) et qui enregistre chaque séquence comme une instance de l’ADTsequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def MatrixParser(filename):\n",
    "    f = open(filename, encoding='utf-8')\n",
    "    matrix = SubstitutionMatrix()\n",
    "    for line in f:\n",
    "        if line[0] not in [\"#\", \" \", \"\\n\"]:\n",
    "            matrix.Append(line[0])\n",
    "            matrix.addRow(line[1:].split())\n",
    "    f.close()\n",
    "    return matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MatrixParser un parser qui peut lire des fichiers qui représentent des matrices de substitution et qui enregistre chaque\n",
    "matrice comme une instance de l’ADT substitution matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "GLOBALsequences = SequencesParser(\"WW-sequence.fasta\")\n",
    "LOCALsequences = SequencesParser(\"protein-sequences.fasta\")\n",
    "pam120 = MatrixParser(\"pam120.txt\")\n",
    "pam60 = MatrixParser(\"pam60.txt\")\n",
    "blosum80 = MatrixParser(\"blosum80.txt\")\n",
    "blosum62 = MatrixParser(\"blosum62.txt\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extraction des données et l'initialisation des différentes matrices de substitutions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ScoringMatrix(ADTmatrix):\n",
    "\n",
    "    def __init__(self, character, mode, sequence1, sequence2, I=None, E=None):\n",
    "        super().__init__(len(sequence1) + 1, len(sequence2) + 1)\n",
    "        self.sequence1 = sequence1\n",
    "        self.sequence2 = sequence2\n",
    "        self.n = len(sequence1) + 1\n",
    "        self.m = len(sequence2) + 1\n",
    "\n",
    "        if character == \"V\":\n",
    "            self.initV()\n",
    "        elif character == \"W\":\n",
    "            self.initW()\n",
    "        else:\n",
    "            self.initS(mode, I, E)\n",
    "\n",
    "    def initS(self, mode, I, E):\n",
    "\n",
    "        for i in range(1, self.m):\n",
    "            gap = - I - (i - 1) * E\n",
    "            if mode == \"LOCAL\":\n",
    "                self.setItem(0, i, max(0, gap))\n",
    "            else:\n",
    "                self.setItem(0, i, gap)\n",
    "\n",
    "        for j in range(1, self.n):\n",
    "            gap = - I - (j - 1) * E\n",
    "            if mode == \"LOCAL\":\n",
    "                self.setItem(j, 0, max(0, gap))\n",
    "            else:\n",
    "                self.setItem(j, 0, gap)\n",
    "\n",
    "    def initV(self):\n",
    "        for i in range(self.m):\n",
    "            self.setItem(0, i, - float(\"inf\"))\n",
    "\n",
    "        for j in range(1, self.n):\n",
    "            self.setItem(j, 0, 0)\n",
    "\n",
    "    def initW(self):\n",
    "        for i in range(1, self.m):\n",
    "            self.setItem(0, i, 0)\n",
    "\n",
    "        for j in range(self.n):\n",
    "            self.setItem(j, 0, - float(\"inf\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ScoringMatrix est une sous-classe de la classe ADTmatrix qui permet de créer trois types d'instance différentes S, V, W. Les matrices V et W ont comme rôle de retenir les pénalités où un trou a déjà été initialisé tandis que S est la matrice utilisée pour construire le parcours. V et W sont initialisé de la façon que chaque matrice respecte ces deux formes respectivement :"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|      |      | S    | E    | Q    | 2    |                            \n",
    "|------|------|------|------|------|------|\n",
    "|      | -inf | -inf | -inf |-inf  |-inf  |\n",
    "|   S  | 0    |      |      |      |      |\n",
    "|   E  | 0    |      |      |      |      |\n",
    "|   Q  | 0    |      |      |      |      |\n",
    "|   1  | 0    |      |      |      |      |\n",
    "\n",
    "                                                      La matrice V\n",
    "                                                      \n",
    "|      |      | S    | E    | Q    | 2    |                            \n",
    "|------|------|------|------|------|------|\n",
    "|      | -inf | 0    | 0    | 0    | 0    |\n",
    "|   S  | -inf |      |      |      |      |\n",
    "|   E  | -inf |      |      |      |      |\n",
    "|   Q  | -inf |      |      |      |      |\n",
    "|   1  | -inf |      |      |      |      |\n",
    "\n",
    "                                                      La matrice W"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Par contre pour initialiser la matrice S faut préciser dans quel alignement il sera utilise local ou global parce que dans l'alignement local les valeurs négatives sont remplacées par des zéros. On remplit chaque case par la valeur du gap affine en utilisant la formule suivante $ g(n_{gap}) = - I - (n_{gap} - 1) E $ ou __I__ référence l'insertion du gap et __E__ l'extension du gap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NeedlemanWunsch(object):\n",
    "    def __init__(self, S, V, W, MATSUB, k, I, E):\n",
    "\n",
    "        self.S = S\n",
    "        self.V = V\n",
    "        self.W = W\n",
    "        self.MATSUB = MATSUB\n",
    "        self.k = k\n",
    "        self.I = I\n",
    "        self.E = E\n",
    "        self.solutionCounter = 0\n",
    "\n",
    "        for i in range(1, self.S.n):\n",
    "            for j in range(1, self.S.m):\n",
    "                self.fill(i, j)\n",
    "        self.findAlignment()\n",
    "\n",
    "\n",
    "    def t(self, i, j):\n",
    "        x = self.MATSUB[self.S.sequence2[j - 1]]\n",
    "        y = self.MATSUB[self.S.sequence1[i - 1]]\n",
    "        return int(self.MATSUB.getItem(x, y))\n",
    "\n",
    "    def fill(self, i, j):\n",
    "\n",
    "        self.V.setItem(i, j, max(self.S.getItem(i - 1, j) - self.I, self.V.getItem(i - 1, j) - self.E))\n",
    "        self.W.setItem(i, j, max(self.S.getItem(i, j - 1) - self.I, self.W.getItem(i, j - 1) - self.E))\n",
    "        self.S.setItem(i, j, max(self.S.getItem(i - 1, j - 1) + self.t(i, j), self.V.getItem(i,j), self.W.getItem(i,j)))        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La classe NeedlemanWunsch est une implémentation de l'algorithme de Needleman-Wunsch qui est un algorithme qui effectue un alignement global entre deux séquences de protéines en utilisant la programmation dynamique. <br>\n",
    "\n",
    "La première étape de l'algorithme est de remplir les matrices de scores S, V et W en utilisant les 3 formules ci-dessous:\n",
    "\n",
    "$$\n",
    "V(i,j) = max\n",
    "    \\left \\{\n",
    "       \\begin{array}{l}\n",
    "          S(i-1,j)-I \\\\\n",
    "          W(i-1,j)-E\n",
    "       \\end{array}\n",
    "       \\right .\n",
    "$$\n",
    "\n",
    "$$\n",
    "W(i,j) = max\n",
    "    \\left \\{\n",
    "       \\begin{array}{l}\n",
    "          S(i,j-1)-I \\\\\n",
    "          W(i,j-1)-E\n",
    "       \\end{array}\n",
    "       \\right .\n",
    "$$\n",
    "\n",
    "$$\n",
    "S(i,j) = max\n",
    "    \\left \\{\n",
    "       \\begin{array}{l}\n",
    "          S(i-1,j-1)+t(i,j) \\\\\n",
    "          V(i,j) \\\\\n",
    "          W(i,j)\n",
    "       \\end{array}\n",
    "       \\right .\n",
    "$$\n",
    "\n",
    "La fonction V renvoie la valeur maximale entre l'insertion d'un nouvel espace dans la séquence horizontale et l'extension des espaces dans la séquence verticale.\n",
    "\n",
    "La fonction W renvoie la valeur maximale entre l'insertion d'un nouvel espace dans la séquence verticale et l'extension des espaces dans la séquence horizontale. \n",
    "\n",
    "La fonction S renvoie le maximum entre le match des deux acides amines en utilisant la valeur stocke dans la matrice de substitution et les résultats obtenus auparavant par les deux fonctions V et W à la même position $(i, j)$ .\n",
    "\n",
    "Dans la classe c'est la méthode fill(i, j) qui s'occupe de remplir les matrices comme c'est expliqué au-dessus et la méthode t(i, j) renvoie la valeur du match entre les deux acides amines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NeedlemanWunsch(NeedlemanWunsch):\n",
    "        \n",
    "    def findAlignment(self, i=None, j=None,seq = \"\",seq2 = \"\",seq3 = \"\"):\n",
    "        if i is None and j is None:\n",
    "            i = self.S.n - 1\n",
    "            j = self.S.m - 1\n",
    "        if i >= 0 and j >= 0 and self.solutionCounter < self.k :\n",
    "            if i == 0 and j == 0:\n",
    "                self.solutionCounter += 1\n",
    "                show(self.solutionCounter,self.S.getItem(self.S.n - 1, self.S.m - 1), seq[::-1], seq2[::-1], seq3[::-1])\n",
    "\n",
    "            else:\n",
    "\n",
    "                if self.S.getItem(i, j) == self.S.getItem(i - 1, j - 1) + self.t(i, j):\n",
    "                    seq += self.S.sequence1[i - 1]\n",
    "                    seq2 += self.S.sequence2[j - 1]\n",
    "                    if self.S.sequence1[i - 1] == self.S.sequence2[j - 1]:\n",
    "                        seq3 += \"|\"\n",
    "                    else:\n",
    "                        seq3 += \":\" if self.t(i, j) >= 0 else \".\"\n",
    "                    self.findAlignment(i - 1, j - 1,seq,seq2,seq3)\n",
    "                    seq = seq[:-1]\n",
    "                    seq2 = seq2[:-1]\n",
    "                    seq3 = seq3[:-1]\n",
    "\n",
    "                if self.S.getItem(i, j) == self.V.getItem(i, j) or j == 0:\n",
    "                    seq += self.S.sequence1[i - 1]\n",
    "                    seq2 += \"-\"\n",
    "                    seq3 += \" \"\n",
    "                    self.findAlignment(i - 1, j,seq,seq2,seq3)\n",
    "                    seq = seq[:-1]\n",
    "                    seq2 = seq2[:-1]\n",
    "                    seq3 = seq3[:-1]\n",
    "\n",
    "                if self.S.getItem(i, j) == self.W.getItem(i, j) or i == 0:\n",
    "                    seq += \"-\"\n",
    "                    seq2 += self.S.sequence2[j - 1]\n",
    "                    seq3 += \" \"\n",
    "                    self.findAlignment(i, j - 1,seq,seq2,seq3)\n",
    "                    seq = seq[:-1]\n",
    "                    seq2 = seq2[:-1]\n",
    "                    seq3 = seq3[:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Après le remplissage des trois matrices, l'algorithme débute la deuxième phase qui consiste à trouver l'alignement optimal entre les deux séquences de protéines. L'idée de cette phase est de trouver tous les alignements possibles qui ont le même meilleur score, pour cela on commence à la position $ (n, m) $ qui correspond a la derniere case de la matrice S qui contient la valeur maximale qu'on peut obtenir pour un alignement entre ces deux séquences puis en essaye de remonter jusqu'à ce qu'on arrive à la valeur 0 qui se trouvait a la position $ (0,0) $ de la matrice du score S. En remontant la décision d'aller à gauche, en haut ou en diagonale depend d'où vient la valeur à cette position. Si la valeur vient du V on insère un gap dans la séquence horizontale s'il vient du W on l'insert dans la séquence verticale sinon la valeur vient du match et donc on ne fait rien. À un moment donné la valeur peut-être obtenue à partir des deux ou les trois matrices mais on donne toujours la priorité au match parce que c'est plus preferable que d'introduire un gap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SmithWaterman(object):\n",
    "    def __init__(self, S, V, W, MATSUB, l, I, E):\n",
    "\n",
    "        self.S = S\n",
    "        self.V = V\n",
    "        self.W = W\n",
    "        self.MATSUB = MATSUB\n",
    "        self.I = I\n",
    "        self.E = E\n",
    "\n",
    "        for i in range(1, self.S.n):\n",
    "            for j in range(1, self.S.m):\n",
    "                self.fill(i, j)\n",
    "\n",
    "        i = 0\n",
    "        while i < l and sum([sum(row) for row in self.S.matrix]):\n",
    "            indexList = []\n",
    "            self.topDown(i+1,indexList)\n",
    "            self.fillZero(indexList)\n",
    "            x, y = indexList[-1]\n",
    "            self.computeAgain(x, y)\n",
    "            i += 1\n",
    "\n",
    "    def t(self, i, j):\n",
    "        x = self.MATSUB[self.S.sequence2[j - 1]]\n",
    "        y = self.MATSUB[self.S.sequence1[i - 1]]\n",
    "        return int(self.MATSUB.getItem(x, y))\n",
    "\n",
    "    def fill(self, i, j):\n",
    "\n",
    "        self.V.setItem(i, j, max(0, self.S.getItem(i - 1, j) - self.I, self.V.getItem(i - 1, j) - self.E))\n",
    "        self.W.setItem(i, j, max(0, self.S.getItem(i, j - 1) - self.I, self.W.getItem(i, j - 1) - self.E))\n",
    "        self.S.setItem(i, j, max(0, self.S.getItem(i - 1, j - 1) + self.t(i, j), self.V.getItem(i,j), self.W.getItem(i,j)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "La classe SmithWaterman est une implémentation de l'algorithme de Smith-Waterman qui est un algorithme qui effectue un alignement local entre deux séquences de protéines en utilisant la programmation dynamique. <br>\n",
    "\n",
    "La première étape de l'algorithme est de remplir les matrices de scores S, V et W en utilisant les 3 formules ci-dessous:\n",
    "\n",
    "$$\n",
    "V(i,j) = max\n",
    "    \\left \\{\n",
    "       \\begin{array}{l}\n",
    "          S(i-1,j)-I \\\\\n",
    "          W(i-1,j)-E \\\\\n",
    "          0\n",
    "       \\end{array}\n",
    "       \\right .\n",
    "$$\n",
    "\n",
    "$$\n",
    "W(i,j) = max\n",
    "    \\left \\{\n",
    "       \\begin{array}{l}\n",
    "          S(i,j-1)-I \\\\\n",
    "          W(i,j-1)-E \\\\\n",
    "          0\n",
    "       \\end{array}\n",
    "       \\right .\n",
    "$$\n",
    "\n",
    "$$\n",
    "S(i,j) = max\n",
    "    \\left \\{\n",
    "       \\begin{array}{l}\n",
    "          S(i-1,j-1)+t(i,j) \\\\\n",
    "          V(i,j) \\\\\n",
    "          W(i,j) \\\\\n",
    "          0\n",
    "       \\end{array}\n",
    "       \\right .\n",
    "$$\n",
    "\n",
    "Les trois fonctions ont le même fonctionnement que celles dans l'alignement global cependant il y a seulement une petite différence par rapport à avant c'est que si la valeur renvoyer et négatifs on la remplace par zéro."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SmithWaterman(SmithWaterman):\n",
    "    def topDown(self, number, indexList):\n",
    "\n",
    "        i, j = self.S.maxValue()\n",
    "        indexList.append((i, j))\n",
    "        score = self.S.getItem(i, j)\n",
    "        seq = \"\"\n",
    "        seq2 = \"\"\n",
    "        seq3 = \"\"\n",
    "        pos = (i,j)\n",
    "        while self.S.getItem(i, j):\n",
    "\n",
    "            value = self.S.getItem(i, j)\n",
    "            pos2 = (i,j)\n",
    "\n",
    "            if value == self.S.getItem(i - 1, j - 1) + self.t(i, j):\n",
    "                seq += self.S.sequence1[i - 1]\n",
    "                seq2 += self.S.sequence2[j - 1]\n",
    "                if self.S.sequence1[i - 1] == self.S.sequence2[j - 1]:\n",
    "                    seq3 += \"|\"\n",
    "                else:\n",
    "                    seq3 += \":\" if self.t(i, j) >= 0 else \".\"\n",
    "                i, j = i - 1, j - 1\n",
    "\n",
    "            elif value == self.V.getItem(i, j):\n",
    "                seq += self.S.sequence1[i - 1]\n",
    "                seq2 += \"-\"\n",
    "                seq3 += \" \"\n",
    "                i = i - 1\n",
    "\n",
    "            elif value == self.W.getItem(i, j):\n",
    "                seq += \"-\"\n",
    "                seq2 += self.S.sequence2[j - 1]\n",
    "                seq3 += \" \"\n",
    "                j = j - 1\n",
    "\n",
    "            indexList.append((i, j))\n",
    "\n",
    "        show(number, score, seq[::-1], seq2[::-1], seq3[::-1],pos,pos2)\n",
    "        indexList.pop()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pour l'alignement local on procède différemment par rapport au global, on ne commence pas à partir de $ (i, j) $ mais à partir de la position de la plus grande valeur de la matrice du score S et s'il y en a plusieurs valeurs maximales on peut prendre n'importe laquelle la seule chose qui va changer c'est juste l'ordre dans lequel notre résultat est obtenu.\n",
    "Donc la methode maxValue() nous renvoie la position de la valeur maximale de la matrice du score et c'est a partir de cette position qu'on va commencer a construire notre alignement on faisant la meme chose qu'on a fait dans l'alignement global sauf que cette fois ci on s'arrete des qu'on rencontre une valeur 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SmithWaterman(SmithWaterman):\n",
    "    def fillZero(self, indexList):\n",
    "\n",
    "        for elem in indexList:\n",
    "            self.S.setItem(elem[0], elem[1], 0)\n",
    "            self.V.setItem(elem[0], elem[1], 0)\n",
    "            self.W.setItem(elem[0], elem[1], 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Des qu'on a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SmithWaterman(SmithWaterman):\n",
    "    def computeAgain(self, x, y):\n",
    "        i, j = x, y\n",
    "        while i < self.S.n:\n",
    "            while j < self.S.m:\n",
    "                if self.S.getItem(i,j) > 0 or self.V.getItem(i,j) > 0 or self.W.getItem(i,j) > 0:\n",
    "                    self.fill(i, j)\n",
    "                j += 1\n",
    "            i += 1\n",
    "            j = y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show(number, score, sequence, sequence2, sequence3, pos=None, pos2=None):\n",
    "    print(\"Soluntion n°: {0} \".format(number))\n",
    "    w = (len(sequence) // 60 + 1) if len(sequence) % 60 != 0 else len(sequence) // 60\n",
    "    for i in range(w):\n",
    "        if pos is None and pos2 is None :\n",
    "            print(sequence[i * 60:(i + 1) * 60] \\\n",
    "                + \"\\n\" + sequence3[i * 60:(i + 1) * 60] \\\n",
    "                + \"\\n\" + sequence2[i * 60:(i + 1) * 60] + \"\\n\")\n",
    "        else :\n",
    "            print(sequence[i * 60:(i + 1) * 60] \\\n",
    "                + \" - from {0} to {1}\".format(pos2[0],pos[0]) \\\n",
    "                + \"\\n\" + sequence3[i * 60:(i + 1) * 60] \\\n",
    "                + \"\\n\" + sequence2[i * 60:(i + 1) * 60] \\\n",
    "                + \" - from {0} to {1}\".format(pos2[1],pos[1]) + \"\\n\")\n",
    "\n",
    "    print(\"Score :\", score)\n",
    "    print(\"Identity percentage :\", round(sequence3.count(\"|\") / len(sequence3) * 100,2), \"%\")\n",
    "    print(\"Similarity rate:\", round((sequence3.count(\"|\") + sequence3.count(\":\")) / len(sequence3) * 100,2), \"%\")\n",
    "    print(\"-\" * 60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 1 and 2\n",
      "Soluntion n°: 1 \n",
      "VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
      ".|||:|||..:...|:.|::||.::.|.|:.|..\n",
      "SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
      "\n",
      "Score : 62\n",
      "Identity percentage : 38.24 %\n",
      "Similarity rate: 61.76 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 1 and 3\n",
      "Soluntion n°: 1 \n",
      "VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
      ".|||:|||....::|:.:::||..:.|.|:|||.\n",
      "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL\n",
      "\n",
      "Score : 76\n",
      "Identity percentage : 41.18 %\n",
      "Similarity rate: 67.65 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 1 and 4\n",
      "Soluntion n°: 1 \n",
      "VPLPAGWEMAKT-SSGQRYFLNHIDQTTTWQDPRK\n",
      "..||:|||:.:: |||:.|::|||.:::.|:.|..\n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "\n",
      "Score : 67\n",
      "Identity percentage : 40.0 %\n",
      "Similarity rate: 71.43 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 1 and 5\n",
      "Soluntion n°: 1 \n",
      "VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
      "....:.|...|:::|:.|:.|...:.:||:.|..\n",
      "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD\n",
      "\n",
      "Score : 30\n",
      "Identity percentage : 23.53 %\n",
      "Similarity rate: 50.0 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 1 and 6\n",
      "Soluntion n°: 1 \n",
      "VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
      ":.....|...|::||:.|:.|...:.:.|..|:.\n",
      "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE\n",
      "\n",
      "Score : 28\n",
      "Identity percentage : 23.53 %\n",
      "Similarity rate: 47.06 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 2 and 3\n",
      "Soluntion n°: 1 \n",
      "SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
      ":|||||||||.::.||::|:||:.:||||..|..\n",
      "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL\n",
      "\n",
      "Score : 113\n",
      "Identity percentage : 55.88 %\n",
      "Similarity rate: 79.41 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 2 and 4\n",
      "Soluntion n°: 1 \n",
      "SPLPPGWEERQD-ILGRTYYVNHESRRTQWKRPTP\n",
      "..||||||.|.: ..||:||.||.:..:||.||:.\n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "\n",
      "Score : 83\n",
      "Identity percentage : 48.57 %\n",
      "Similarity rate: 62.86 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "SPLPPGWEERQDI-LGRTYYVNHESRRTQWKRPTP\n",
      "..||||||.|.:. .||:||.||.:..:||.||:.\n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "\n",
      "Score : 83\n",
      "Identity percentage : 48.57 %\n",
      "Similarity rate: 62.86 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "SPLPPGWEERQDIL-GRTYYVNHESRRTQWKRPTP\n",
      "..||||||.|.:.. ||:||.||.:..:||.||:.\n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "\n",
      "Score : 83\n",
      "Identity percentage : 48.57 %\n",
      "Similarity rate: 62.86 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 2 and 5\n",
      "Soluntion n°: 1 \n",
      "SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
      "|...:.|.|:::..|||||.|.|::::.|.:|..\n",
      "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD\n",
      "\n",
      "Score : 55\n",
      "Identity percentage : 35.29 %\n",
      "Similarity rate: 61.76 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 2 and 6\n",
      "Soluntion n°: 1 \n",
      "SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
      "......|.|.::..|:.||.|.:::.::|.:|..\n",
      "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE\n",
      "\n",
      "Score : 27\n",
      "Identity percentage : 23.53 %\n",
      "Similarity rate: 50.0 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 3 and 4\n",
      "Soluntion n°: 1 \n",
      "GPLPPGWEER-THTDGRIFYINHNIKRTQWEDPRL\n",
      "..||||||.| ::::||::|:||.::.:|||.|..\n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "\n",
      "Score : 89\n",
      "Identity percentage : 45.71 %\n",
      "Similarity rate: 74.29 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 3 and 5\n",
      "Soluntion n°: 1 \n",
      "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL\n",
      ":...:.|.|:...|||::|.|.::|::.||.|..\n",
      "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD\n",
      "\n",
      "Score : 51\n",
      "Identity percentage : 32.35 %\n",
      "Similarity rate: 58.82 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 3 and 6\n",
      "Soluntion n°: 1 \n",
      "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL\n",
      "......|.|....:|:.:|.|.::|.::|:.|:.\n",
      "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE\n",
      "\n",
      "Score : 23\n",
      "Identity percentage : 23.53 %\n",
      "Similarity rate: 50.0 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 4 and 5\n",
      "Soluntion n°: 1 \n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "....:.|..::| ::||:||:|..|:.|.||:|::\n",
      "SGAKSMWTEHKS-PDGRTYYYNTETKQSTWEKPDD\n",
      "\n",
      "Score : 49\n",
      "Identity percentage : 34.29 %\n",
      "Similarity rate: 65.71 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 4 and 6\n",
      "Soluntion n°: 1 \n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "......| |....:||:.||:|..|::|:|::|..\n",
      "LLSKCPW-KEYKSDSGKPYYYNSQTKESRWAKPKE\n",
      "\n",
      "Score : 30\n",
      "Identity percentage : 31.43 %\n",
      "Similarity rate: 54.29 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
      "......|...:| :||:.||:|..|::|:|::|..\n",
      "LLSKCPWKEYKS-DSGKPYYYNSQTKESRWAKPKE\n",
      "\n",
      "Score : 30\n",
      "Identity percentage : 31.43 %\n",
      "Similarity rate: 57.14 %\n",
      "------------------------------------------------------------\n",
      "\n",
      "\n",
      "\n",
      "Alignment between sequence 5 and 6\n",
      "Soluntion n°: 1 \n",
      "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD\n",
      "..:|:.|.|.||.:|:.||||::||:|.|:||.:\n",
      "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE\n",
      "\n",
      "Score : 89\n",
      "Identity percentage : 47.06 %\n",
      "Similarity rate: 73.53 %\n",
      "------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "I = 12\n",
    "E = 2\n",
    "k = 5\n",
    "for i in range(len(GLOBALsequences)):\n",
    "    for j in range(i+1,len(GLOBALsequences)) :\n",
    "        print(\"\\n\" * 2)\n",
    "        print(\"Alignment between sequence {0} and {1}\".format(i+1,j+1))\n",
    "        S = ScoringMatrix(\"S\",\"GLOBAL\",GLOBALsequences[i],GLOBALsequences[j],I,E)\n",
    "        V = ScoringMatrix(\"V\",\"GLOBAL\",GLOBALsequences[i],GLOBALsequences[j])\n",
    "        W = ScoringMatrix(\"W\",\"GLOBAL\",GLOBALsequences[i],GLOBALsequences[j])\n",
    "        NeedlemanWunsch(S,V,W,pam120,k,I,E)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment between sequence 1 and 2\n",
      "Soluntion n°: 1 \n",
      "LPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRKAMLSQMNVTAPTSPPVQQNMMNSASGPL - from 173 to 264\n",
      "||.|||:..:.:|:.:|::|..:||||:|||      :::.|.....::.:..|:. |||\n",
      "LPKGWEVRHAPNGRPFFIDHNTKTTTWEDPR------LKIPAHLRGKTSLDTSNDL-GPL - from 842 to 926\n",
      "\n",
      "PDGWEQAMTQDGEIYYINHKNKTTSWLDPRLD - from 173 to 264\n",
      "|.|||:....||:|:||||:.|.|:|.||||:\n",
      "PPGWEERTHTDGRIFYINHNIKRTQWEDPRLE - from 842 to 926\n",
      "\n",
      "Score : 194\n",
      "Identity percentage : 41.3 %\n",
      "Similarity rate: 65.22 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "IPDDVPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDP--------RKAMLSQMNVT----A - from 167 to 263\n",
      ":|.:..||.|||..:.:.|:.|:::|.::||||..|        .:...||.::.    |\n",
      "LPTSSGLPPGWEEKQDERGRSYYVDHNSRTTTWTKPTVQATVETSQLTSSQSSAGPQSQA - from 763 to 873\n",
      "\n",
      "PTSPPVQQ-NMMNSA-SGPLPDGWEQAMTQDGEIYYINHKNKTTSWLDPRL - from 167 to 263\n",
      ".||...|| ::.::. :|.||.|||...:.:|:.::|:|::|||:|.||||\n",
      "STSDSGQQVTQPSEIEQGFLPKGWEVRHAPNGRPFFIDHNTKTTTWEDPRL - from 763 to 873\n",
      "\n",
      "Score : 149\n",
      "Identity percentage : 34.23 %\n",
      "Similarity rate: 58.56 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "HLR-QSSFEIPDDV-PLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRKAMLSQMNVTAPT - from 159 to 218\n",
      "||| ::|::..:|: |||.|||....::|:.:::||..:.|.|:|||   |:::.:|:|:\n",
      "HLRGKTSLDTSNDLGPLPPGWEERTHTDGRIFYINHNIKRTQWEDPR---LENVAITGPA - from 878 to 936\n",
      "\n",
      "SP - from 159 to 218\n",
      ".|\n",
      "VP - from 878 to 936\n",
      "\n",
      "Score : 103\n",
      "Identity percentage : 37.1 %\n",
      "Similarity rate: 69.35 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 1 and 3\n",
      "Soluntion n°: 1 \n",
      "RQSSFEIPDDVPLPA-------GWEMAKTSSGQRYFLNHIDQTTTWQDPRK - from 161 to 204\n",
      ":||::|.|||:..||       .|:..|::||:.|:.|...:.:.|..|::\n",
      "KQSTWEKPDDLKTPAEQLLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE - from 164 to 214\n",
      "\n",
      "Score : 70\n",
      "Identity percentage : 31.37 %\n",
      "Similarity rate: 56.86 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "MLSQMN-VTA-PTSPPVQQNMMNSASGPL-PDG-WEQAMTQDGEIYYINHKNKTTSWLDP - from 206 to 261\n",
      "|:|:|: ::. |:.||..::|..:|:... ..: |.:..:.||:.||.|.::|.::|..|\n",
      "MMSHMSQASMQPALPPGVNSMDVAAGTASGAKSMWTEHKSPDGRTYYYNTETKQSTWEKP - from 112 to 171\n",
      "\n",
      "Score : 63\n",
      "Identity percentage : 28.33 %\n",
      "Similarity rate: 58.33 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "WEQAMTQDGEIYYINHKNKTTSWLDPR - from 236 to 262\n",
      "|::..:::|:.||.|.::|.:.|..|:\n",
      "WKEYKSDSGKPYYYNSQTKESRWAKPK - from 187 to 213\n",
      "\n",
      "Score : 57\n",
      "Identity percentage : 29.63 %\n",
      "Similarity rate: 66.67 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 1 and 4\n",
      "Soluntion n°: 1 \n",
      "DDVPLPAGWEMAKT-SSGQRYFLNHIDQTTTWQDP - from 169 to 202\n",
      "|:..||.|||...: |||:.|::|||.:::.|:.|\n",
      "DEEKLPPGWEKRMSRSSGRVYYFNHITNASQWERP - from 3 to 37\n",
      "\n",
      "Score : 80\n",
      "Identity percentage : 42.86 %\n",
      "Similarity rate: 68.57 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "LPDGWEQAMTQ-DGEIYYINHKNKTTSWLDP - from 232 to 261\n",
      "||.|||:.|:: :|::||:||.:::::|..|\n",
      "LPPGWEKRMSRSSGRVYYFNHITNASQWERP - from 7 to 37\n",
      "\n",
      "Score : 75\n",
      "Identity percentage : 41.94 %\n",
      "Similarity rate: 80.65 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "TGVVSGPAATPTAQHL - from 145 to 160\n",
      "||.:|||:.|.::.|:\n",
      "TGEMSGPVFTDSGIHI - from 143 to 158\n",
      "\n",
      "Score : 37\n",
      "Identity percentage : 43.75 %\n",
      "Similarity rate: 75.0 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 1 and 5\n",
      "Soluntion n°: 1 \n",
      "VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDP - from 171 to 202\n",
      "|.||.||:...:.:|:||::|...:.|||:.|\n",
      "VILPPGWQSYLSPQGRRYYVNTTTNETTWERP - from 77 to 108\n",
      "\n",
      "Score : 77\n",
      "Identity percentage : 40.62 %\n",
      "Similarity rate: 65.62 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "LPDGWEQAMTQDGEIYYINHKNKTTSWLDPRLDP - from 232 to 265\n",
      "||.||::.::.:|:.||:|..::.|:|..|..:|\n",
      "LPPGWQSYLSPQGRRYYVNTTTNETTWERPSSSP - from 79 to 112\n",
      "\n",
      "Score : 74\n",
      "Identity percentage : 35.29 %\n",
      "Similarity rate: 67.65 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "QSPQG-GVMGGSNSNQQQQMRLQQLQMEKERLRLKQQELLRQAMRNINPSTANSPKCQE - from 288 to 345\n",
      "::||| |:::|.:...|:|::.:|:|.|.:::..::.::.::..:|:...:.||...||\n",
      "KDPQGNGTVAGFELLLQKQLKGKQMQKEMSEFIRERIKIEEDYAKNLAKLSQNSLASQE - from 209 to 267\n",
      "\n",
      "Score : 49\n",
      "Identity percentage : 25.42 %\n",
      "Similarity rate: 64.41 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 1 and 6\n",
      "Soluntion n°: 1 \n",
      "LPAGWEMAKTSSGQRYFLNHIDQTTTWQDPR------------KAMLSQMNVTAPTSPPV - from 173 to 324\n",
      "||.|:|:..|.:||.|||:....::||:|||            .|:|.::.:.:.||.|.\n",
      "LPEGYEQRTTVQGQVYFLHTQTGVSTWHDPRIPSPSGTIPGGDAAFLYEFLLQGHTSEPR - from 236 to 395\n",
      "\n",
      "QQNMMN-SASGPLPDGWEQAMTQDGEIYYINHKNKTTSWLDPRLDPRFAMNQRISQSAPV - from 173 to 324\n",
      ":.|.:| :..||||.|||...|.:|:||:::|:|:||::.||||      ::.:::::.:\n",
      "DLNSVNCDELGPLPPGWEVRSTVSGRIYFVDHNNRTTQFTDPRL------HHIMNHQCQL - from 236 to 395\n",
      "\n",
      "KQPPPLAPQSPQGGVMGGSNSNQQQQMR-LQQLQMEKERLRLKQQE - from 173 to 324\n",
      "|:|....|...:|::...:...|:.:.. :|:|::.:::|.|:|.:\n",
      "KEPSQPLPLPSEGSLEDEELPAQRYERDLVQKLKVLRHELSLQQPQ - from 236 to 395\n",
      "\n",
      "Score : 180\n",
      "Identity percentage : 30.72 %\n",
      "Similarity rate: 59.64 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "DDV-PLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPR--KAMLSQMNVTAPTSP - from 169 to 218\n",
      "|:: |||.|||:.:|.||:.||::|.::||.:.|||  ..|..|.::..|::|\n",
      "DELGPLPPGWEVRSTVSGRIYFVDHNNRTTQFTDPRLHHIMNHQCQLKEPSQP - from 303 to 355\n",
      "\n",
      "Score : 105\n",
      "Identity percentage : 41.51 %\n",
      "Similarity rate: 67.92 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "QDPRKAMLSQMNVTAPTSPPVQQNMMNSASGP-LPDGWEQAMTQDGEIYYINHKNKTTSW - from 200 to 263\n",
      "||.|.......|..:..|..:.||..::.::| ||:|:||..|.:|::|:::.::.:::|\n",
      "QDQRLQAQRLRNPDVRGSLQTPQNRPHGHQSPELPEGYEQRTTVQGQVYFLHTQTGVSTW - from 203 to 267\n",
      "\n",
      "LDPRL - from 200 to 263\n",
      ".|||:\n",
      "HDPRI - from 203 to 267\n",
      "\n",
      "Score : 88\n",
      "Identity percentage : 30.77 %\n",
      "Similarity rate: 61.54 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 2 and 3\n",
      "Soluntion n°: 1 \n",
      "WEVRHAPNGRPFFIDHNTKTTTWEDP-RLKIPAHLRGKTSLDTSNDLGPLPPGWEERTHT - from 846 to 924\n",
      "|..:.:|:||.::.:.:||.:|||.| .||.||           ::|....| |:|....\n",
      "WTEHKSPDGRTYYYNTETKQSTWEKPDDLKTPA-----------EQLLSKCP-WKEYKSD - from 146 to 213\n",
      "\n",
      "DGRIFYINHNIKRTQWEDPR - from 846 to 924\n",
      ":|:.:|.|.:.|:::|..|:\n",
      "SGKPYYYNSQTKESRWAKPK - from 146 to 213\n",
      "\n",
      "Score : 88\n",
      "Identity percentage : 30.0 %\n",
      "Similarity rate: 53.75 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "PASSSNHSSRRGSLQAYTFEEQPTLPVLLPTSSGLPPGWEEKQDERGRSYYVDHNSRTTT - from 735 to 798\n",
      "|:...:|.| ::|:|........::.|...|:||....|.|.::..||:||.:.:::.:|\n",
      "PGMMMSHMS-QASMQPALPPGVNSMDVAAGTASGAKSMWTEHKSPDGRTYYYNTETKQST - from 109 to 171\n",
      "\n",
      "WTKP - from 735 to 798\n",
      "|.||\n",
      "WEKP - from 109 to 171\n",
      "\n",
      "Score : 79\n",
      "Identity percentage : 29.69 %\n",
      "Similarity rate: 54.69 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "WEERQDILGRTYYVNHESRRTQWKRP - from 616 to 641\n",
      "|.|:::..|||||.|.|::::.|::|\n",
      "WTEHKSPDGRTYYYNTETKQSTWEKP - from 146 to 171\n",
      "\n",
      "Score : 72\n",
      "Identity percentage : 42.31 %\n",
      "Similarity rate: 76.92 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 2 and 4\n",
      "Soluntion n°: 1 \n",
      "LPPGWEERQD-ILGRTYYVNHESRRTQWKRPT - from 612 to 642\n",
      "||||||:|:: ..||:||.||.::.:||:||:\n",
      "LPPGWEKRMSRSSGRVYYFNHITNASQWERPS - from 7 to 38\n",
      "\n",
      "Score : 96\n",
      "Identity percentage : 53.12 %\n",
      "Similarity rate: 81.25 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "LPPGWEER-THTDGRIFYINHNIKRTQWEDP - from 894 to 923\n",
      "||||||:| ::::||::|:||..:.:|||.|\n",
      "LPPGWEKRMSRSSGRVYYFNHITNASQWERP - from 7 to 37\n",
      "\n",
      "Score : 96\n",
      "Identity percentage : 51.61 %\n",
      "Similarity rate: 83.87 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "LPPGWEEKQD-ERGRSYYVDHNSRTTTWTKPT - from 769 to 799\n",
      "||||||:::: :.||.||.:|.::::.|.:|:\n",
      "LPPGWEKRMSRSSGRVYYFNHITNASQWERPS - from 7 to 38\n",
      "\n",
      "Score : 79\n",
      "Identity percentage : 40.62 %\n",
      "Similarity rate: 78.12 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 2 and 5\n",
      "Soluntion n°: 1 \n",
      "LPPGWEEKQDERGRSYYVDHNSRTTTWTKPTVQATVETSQLTSSQSSAGP - from 769 to 818\n",
      "|||||::..:.:||.|||:.:::.|||.:|:.:..:.:|. .|.:||..|\n",
      "LPPGWQSYLSPQGRRYYVNTTTNETTWERPSSSPGIPASP-GSHRSSLPP - from 79 to 127\n",
      "\n",
      "Score : 92\n",
      "Identity percentage : 38.0 %\n",
      "Similarity rate: 66.0 %\n",
      "------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Soluntion n°: 2 \n",
      "LPPGWEERQDILGRTYYVNHESRRTQWKRPT - from 612 to 642\n",
      "|||||::..:..||.||||..:::|.|:||:\n",
      "LPPGWQSYLSPQGRRYYVNTTTNETTWERPS - from 79 to 109\n",
      "\n",
      "Score : 89\n",
      "Identity percentage : 48.39 %\n",
      "Similarity rate: 74.19 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "EIEQGFLPKGWEVRHAPNGRPFFIDHNTKTTTWEDP - from 836 to 871\n",
      "|.:..:||.||:...:|:||.::::.:|:.||||.|\n",
      "ESQTVILPPGWQSYLSPQGRRYYVNTTTNETTWERP - from 73 to 108\n",
      "\n",
      "Score : 83\n",
      "Identity percentage : 38.89 %\n",
      "Similarity rate: 69.44 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 2 and 6\n",
      "Soluntion n°: 1 \n",
      "LPKGWEVRHAPNGRPFFIDHNTKTTTWEDPRL-----KIP----A----H-LRGKTS--- - from 842 to 1315\n",
      "||:|:|.|.:.:|:.:|:..:|.::||:|||:     .||    |    . |:|.||   \n",
      "LPEGYEQRTTVQGQVYFLHTQTGVSTWHDPRIPSPSGTIPGGDAAFLYEFLLQGHTSEPR - from 236 to 754\n",
      "\n",
      "-LDTSN--DLGPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRLENV------------- - from 842 to 1315\n",
      " |::.|  :|||||||||.|:.::|||::::||.:.||:.||||:::             \n",
      "DLNSVNCDELGPLPPGWEVRSTVSGRIYFVDHNNRTTQFTDPRLHHIMNHQCQLKEPSQP - from 236 to 754\n",
      "\n",
      "-------AITGPAVP---YSRDYKRKYEFFRRKLKKQNDIPNKFEMKLRRATVLEDSYRR - from 842 to 1315\n",
      "       ::....:|   |:||..:|.:.:|::|:.|:...:..::::.|..::|:|||:\n",
      "LPLPSEGSLEDEELPAQRYERDLVQKLKVLRHELSLQQPQAGHCRIEVSREEIFEESYRQ - from 236 to 754\n",
      "\n",
      "IMGVKRADFLKARLWIEFDGEKGLDYGGVAREWFFLISKEMFNPYYGLFEYSATDNYTLQ - from 842 to 1315\n",
      "||.::..| ||.||.::|.||:|||||||||||::|:..||:|||||||:||:.:.|.||\n",
      "IMKMRPKD-LKKRLMVKFRGEEGLDYGGVAREWLYLLCHEMLNPYYGLFQYSTDNIYMLQ - from 236 to 754\n",
      "\n",
      "INPNSGLCNEDHLSYFKFIGRVAGMAVYHGKLLDGFFIRPFYKMMLHKPITLHDMESVDS - from 842 to 1315\n",
      "|||:|:: |.||||||.|:||:.|:||:||..::|.|..||||::|.|||.|.|:||||.\n",
      "INPDSSI-NPDHLSYFHFVGRIMGLAVFHGHYINGGFTVPFYKQLLGKPIQLSDLESVDP - from 236 to 754\n",
      "\n",
      "EYYNSLRWILENDPTE-LDLRFIIDEELFGQTHQHELKNGGSEIVVTNKNKKEYIYLVIQ - from 842 to 1315\n",
      "|.::||.||||||.|. ||..|.::::.||:..|||||.:|.::.||::|||||:.|.::\n",
      "ELHKSLVWILENDITPVLDHTFCVEHNAFGRILQHELKPNGRNVPVTEENKKEYVRLYVN - from 236 to 754\n",
      "\n",
      "WRFVNRIQKQMAAFKEGFFELIPQDLIKIFDENELELLMCGLGDVDVNDWREHTKYKNGY - from 842 to 1315\n",
      "|||::.|:.|:.|:::||.|||||.|:|.||::||||::.||..:|:|||:::|:.|: .\n",
      "WRFMRGIEAQFLALQKGFNELIPQHLLKPFDQKELELIIGGLDKIDLNDWKSNTRLKH-C - from 236 to 754\n",
      "\n",
      "SANHQVIQWFWKAVLMMDSEKRIRLLQFVTGTSRVPMNGFAELYGS-N--GPQSFTVEQW - from 842 to 1315\n",
      ".|:.::::|||:||..:|:|:|.||||||||::|||::||..|.|| :  ||:.||::..\n",
      "VADSNIVRWFWQAVETFDEERRARLLQFVTGSTRVPLQGFKALQGSTGAAGPRLFTIHLI - from 236 to 754\n",
      "\n",
      "-GTPEKLPRAHTCFNRLDLPPYESFEELWDKLQMAIENTQGF - from 842 to 1315\n",
      " ::.::||:|||||||:|:|||||:|:|::||..|:|:|.||\n",
      "DANTDNLPKAHTCFNRIDIPPYESYEKLYEKLLTAVEETCGF - from 236 to 754\n",
      "\n",
      "Score : 1113\n",
      "Identity percentage : 45.59 %\n",
      "Similarity rate: 71.65 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "TDAENGNIQLQAQRAFTTRRQISEETESVDNRESSENWEIIREDEATMYSNQAFPSPPPS - from 648 to 873\n",
      "|||.:|:|.::.|    ||.:|:...:.||.|...||...:.||::.....::| ...|:\n",
      "TDAVRGQIVVSLQ----TRDRIGTGGSVVDCRGLLENEGTVYEDSGPGRPLSCF-MEEPA - from 126 to 339\n",
      "\n",
      "SNLDVPTHLAEELNARLTIFGNSAVSQPASSSNHSSRRGSLQAYTFEEQPTLPVLLPTSS - from 648 to 873\n",
      "...|.....|...|:|::...::.....|:..::.:.|||||  |   ..:.|. ...|.\n",
      "PYTDSTGAAAGGGNCRFVESPSQDQRLQAQRLRNPDVRGSLQ--T---PQNRPH-GHQSP - from 126 to 339\n",
      "\n",
      "GLPPGWEEKQDERGRSYYVDHNSRTTTWTKPTVQATVETSQLTSSQSSAGPQSQASTSDS - from 648 to 873\n",
      ".||.|:|::...:|:.|::..::.::||..|.:.:...|....::........|:.||:.\n",
      "ELPEGYEQRTTVQGQVYFLHTQTGVSTWHDPRIPSPSGTIPGGDAAFLYEFLLQGHTSEP - from 126 to 339\n",
      "\n",
      "GQQVTQPSEIEQGFLPKGWEVRHAPNGRPFFIDHNTKTTTWEDPRL - from 648 to 873\n",
      " :::::.:..|.|.||.|||||.:.:||.:|:|||::||.:.||||\n",
      "-RDLNSVNCDELGPLPPGWEVRSTVSGRIYFVDHNNRTTQFTDPRL - from 126 to 339\n",
      "\n",
      "Score : 153\n",
      "Identity percentage : 28.76 %\n",
      "Similarity rate: 52.21 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "LQAYTFEEQPTLPVLLPTSSGLPPGWEEKQDERGRSYYVDHNSRTTTWTKP - from 748 to 798\n",
      "||::|.|.:....|.....:.||||||.::...||.|:||||:|||.:|.|\n",
      "LQGHTSEPRDLNSVNCDELGPLPPGWEVRSTVSGRIYFVDHNNRTTQFTDP - from 287 to 337\n",
      "\n",
      "Score : 113\n",
      "Identity percentage : 45.1 %\n",
      "Similarity rate: 62.75 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 3 and 4\n",
      "Soluntion n°: 1 \n",
      "KEYKSDSGKPYYYNSQTKESRWAKPKELEDLEGYQNTIVAGSLITKSNL--HAMIKAEES - from 188 to 254\n",
      "|:.:.:||:.||:|..|:.|:|.:|:..::..|.::....:.:..:..|  |::.:..:|\n",
      "KRMSRSSGRVYYFNHITNASQWERPSGNSSSGGKNGQGEPARVRCSHLLVKHSQSRRPSS - from 13 to 81\n",
      "\n",
      "SKQEECTTT - from 188 to 254\n",
      ".:||:.|.|\n",
      "WRQEKITRT - from 13 to 81\n",
      "\n",
      "Score : 63\n",
      "Identity percentage : 26.09 %\n",
      "Similarity rate: 57.97 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "WTEHKS-PDGRTYYYNTETKQSTWEKP - from 146 to 171\n",
      "|.::.| .:||:||:|..|:.|.||:|\n",
      "WEKRMSRSSGRVYYFNHITNASQWERP - from 11 to 37\n",
      "\n",
      "Score : 62\n",
      "Identity percentage : 44.44 %\n",
      "Similarity rate: 70.37 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "SRWAKPKELEDLE-GYQNTIVAGSLITKSNLHAMIKAE - from 207 to 243\n",
      "||....|.:||.: :.::..::|.::|.|::|.::::|\n",
      "SRGQMQKPFEDASFALRTGEMSGPVFTDSGIHIILRTE - from 126 to 163\n",
      "\n",
      "Score : 36\n",
      "Identity percentage : 26.32 %\n",
      "Similarity rate: 65.79 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 3 and 5\n",
      "Soluntion n°: 1 \n",
      "PMGGPPMGQMPGMMSSVMPGMMMSHMSQASMQPALPPGVNSMDVAAGTASGAKSMWTEHK - from 91 to 171\n",
      "|.||...|:....:.:.:|:..::.:::.:|.|. |||.:|:.|   ..  ..: |.::.\n",
      "PDGGWWEGEKEDGLRGWFPASYVQLLEKPGMVPP-PPGEESQTV---IL--PPG-WQSYL - from 35 to 108\n",
      "\n",
      "SPDGRTYYYNTETKQSTWEKP - from 91 to 171\n",
      "||:||.||.||.|:::|||:|\n",
      "SPQGRRYYVNTTTNETTWERP - from 35 to 108\n",
      "\n",
      "Score : 98\n",
      "Identity percentage : 32.1 %\n",
      "Similarity rate: 58.02 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "WTEHKSPDGRTYYYNTETKQSTWEKPDDL-KTP---AEQLLSKCPWKEYKSDSGKPYYYN - from 146 to 212\n",
      "|.|.::.||...::.::..| ..|||..: ..|   ::.::....|::|.|.:|:.||.|\n",
      "WWEGEKEDGLRGWFPASYVQ-LLEKPGMVPPPPGEESQTVILPPGWQSYLSPQGRRYYVN - from 39 to 108\n",
      "\n",
      "SQTKESRWAKP - from 146 to 212\n",
      ":.|:|:.|.:|\n",
      "TTTNETTWERP - from 39 to 108\n",
      "\n",
      "Score : 79\n",
      "Identity percentage : 28.17 %\n",
      "Similarity rate: 54.93 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "PMGGPPMGQMPGMMSSVMPGMMMSHMSQASMQPALPPGVNSMDVAAGTASGAKSMWTEHK - from 91 to 187\n",
      "|.::|.:...||...|.:|..:.::  :||..||.||.:..|:|..:|  |. |:....:\n",
      "PSSSPGIPASPGSHRSSLPPTVNGY--HASGTPAHPPETAHMSVRKST--GD-SQNLGSS - from 108 to 198\n",
      "\n",
      "SPDGRTYYYNTETKQSTWEKPDDLKTPAEQLLSKCPW - from 91 to 187\n",
      "||:.:....||.|.:.:.....|.. |.:|||:...|\n",
      "SPSKKQSKENTITINCVTFPHPDTM-PEQQLLKPTEW - from 108 to 198\n",
      "\n",
      "Score : 59\n",
      "Identity percentage : 28.87 %\n",
      "Similarity rate: 48.45 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 3 and 6\n",
      "Soluntion n°: 1 \n",
      "WTEHKSPDGRTYYYNTETKQSTWEKP - from 146 to 171\n",
      ":.::.:.:|::|:.:|:|..|||:.|\n",
      "YEQRTTVQGQVYFLHTQTGVSTWHDP - from 240 to 265\n",
      "\n",
      "Score : 52\n",
      "Identity percentage : 30.77 %\n",
      "Similarity rate: 73.08 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "GKPYYYNSQTKESRWAKPK - from 195 to 213\n",
      "|:.|:.::||..|.|..|:\n",
      "GQVYFLHTQTGVSTWHDPR - from 248 to 266\n",
      "\n",
      "Score : 42\n",
      "Identity percentage : 36.84 %\n",
      "Similarity rate: 63.16 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "PDGRTYYYNTETKQSTWEKPDDLKT-PAEQLLSKCP-WKEYKSDSGKPYYYNSQTKESRW - from 152 to 213\n",
      "|.|.:.:......|:..::|.||:: .:::|....| |:..::.||:.|:.:.:::.:::\n",
      "PGGDAAFLYEFLLQGHTSEPRDLNSVNCDELGPLPPGWEVRSTVSGRIYFVDHNNRTTQF - from 275 to 338\n",
      "\n",
      "AKPK - from 152 to 213\n",
      ":.|:\n",
      "TDPR - from 275 to 338\n",
      "\n",
      "Score : 39\n",
      "Identity percentage : 20.31 %\n",
      "Similarity rate: 57.81 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 4 and 5\n",
      "Soluntion n°: 1 \n",
      "LPPGWEKRMSRSSGRVYYFNHITNASQWERP--S-GNSSSGGKNGQGEPARVRCSH - from 7 to 59\n",
      "|||||::.:| .:||.||.|..||.:.||||  | |..:|.|::.::.|..|:..|\n",
      "LPPGWQSYLS-PQGRRYYVNTTTNETTWERPSSSPGIPASPGSHRSSLPPTVNGYH - from 79 to 133\n",
      "\n",
      "Score : 97\n",
      "Identity percentage : 42.86 %\n",
      "Similarity rate: 62.5 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "EKITRTKEEALELINGYIQKIKSGEEDFESLASQFSDCSSAKARGDLGAFSRGQMQKPFE - from 76 to 136\n",
      ":|..:.|:...|:.:...::||..|:..::||:..::..:::.:|:||. :.:|::|.:.\n",
      "QKQLKGKQMQKEMSEFIRERIKIEEDYAKNLAKLSQNSLASQEEGSLGE-AWAQVKKSLA - from 225 to 284\n",
      "\n",
      "D - from 76 to 136\n",
      "|\n",
      "D - from 225 to 284\n",
      "\n",
      "Score : 37\n",
      "Identity percentage : 22.95 %\n",
      "Similarity rate: 60.66 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "SSWRQEKITRTKE-EALEL - from 71 to 88\n",
      "|:|.:|.:|.|.| |.||:\n",
      "SKWFEEMVTTTLELERLEV - from 395 to 413\n",
      "\n",
      "Score : 35\n",
      "Identity percentage : 47.37 %\n",
      "Similarity rate: 68.42 %\n",
      "------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment between sequence 4 and 6\n",
      "Soluntion n°: 1 \n",
      "LPPGWEKRMSRSSGRVYYFNHITNASQWERP - from 7 to 37\n",
      "||||||.| |..|||:|:.:|.::::|:..|\n",
      "LPPGWEVR-STVSGRIYFVDHNNRTTQFTDP - from 308 to 337\n",
      "\n",
      "Score : 76\n",
      "Identity percentage : 48.39 %\n",
      "Similarity rate: 74.19 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "KLPPGWEKRMSRSSGRVYYFNHITNASQWERPSGNSSSG - from 6 to 44\n",
      ":||.|:|:| :..:|:||:::..|::|.|:.|...|.||\n",
      "ELPEGYEQR-TTVQGQVYFLHTQTGVSTWHDPRIPSPSG - from 235 to 272\n",
      "\n",
      "Score : 63\n",
      "Identity percentage : 38.46 %\n",
      "Similarity rate: 69.23 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "EEALELINGYIQKI - from 83 to 96\n",
      "::.||||.|.::||\n",
      "QKELELIIGGLDKI - from 625 to 638\n",
      "\n",
      "Score : 30\n",
      "Identity percentage : 50.0 %\n",
      "Similarity rate: 78.57 %\n",
      "------------------------------------------------------------\n",
      "Alignment between sequence 5 and 6\n",
      "Soluntion n°: 1 \n",
      "QTVILPPGWQSYLSPQGRRYYVNTTTNETTWERPSSSPGIPASPGSHRSSL - from 75 to 125\n",
      "|:..||.|:::..:.||:.|:::|.|:.:||:.|.......:.||:..:.|\n",
      "QSPELPEGYEQRTTVQGQVYFLHTQTGVSTWHDPRIPSPSGTIPGGDAAFL - from 232 to 282\n",
      "\n",
      "Score : 68\n",
      "Identity percentage : 29.41 %\n",
      "Similarity rate: 58.82 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 2 \n",
      "LPPGWQSYLSPQGRRYYVNTTTNETTWERP - from 79 to 108\n",
      "|||||:...:.:||.|:|:.:::.|.:..|\n",
      "LPPGWEVRSTVSGRIYFVDHNNRTTQFTDP - from 308 to 337\n",
      "\n",
      "Score : 59\n",
      "Identity percentage : 36.67 %\n",
      "Similarity rate: 66.67 %\n",
      "------------------------------------------------------------\n",
      "Soluntion n°: 3 \n",
      "WADKKDPQGNGTVAG - from 205 to 219\n",
      "|.|.:.|:.:||:.|\n",
      "WHDPRIPSPSGTIPG - from 262 to 276\n",
      "\n",
      "Score : 38\n",
      "Identity percentage : 40.0 %\n",
      "Similarity rate: 66.67 %\n",
      "------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "I = 12\n",
    "E = 2\n",
    "l = 3\n",
    "for i in range(len(LOCALsequences)):\n",
    "    for j in range(i+1,len(LOCALsequences)) :\n",
    "            print(\"Alignment between sequence {0} and {1}\".format(i+1,j+1))\n",
    "            S = ScoringMatrix(\"S\",\"LOCAL\",LOCALsequences[i],LOCALsequences[j],I,E)\n",
    "            V = ScoringMatrix(\"V\",\"LOCAL\",LOCALsequences[i],LOCALsequences[j])\n",
    "            W = ScoringMatrix(\"W\",\"LOCAL\",LOCALsequences[i],LOCALsequences[j])\n",
    "            SmithWaterman(S, V, W, blosum62, l, I, E)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Références\n",
    "[1] https://fr.wikipedia.org/wiki/Alignement_de_s%C3%A9quences <br>\n",
    "[2] https://fr.wikipedia.org/wiki/Programmation_dynamique  <br>\n",
    "[3] https://fr.wikipedia.org/wiki/Matrice_de_similarit%C3%A9 <br>\n",
    "<br>\n",
    "<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
